### DISCUSSION TITLE
June's feedback on Recipe 1

### DISCUSSION THREAD

**Discussion by yjunechoe** [EvaMaeRey/easy-geom-recipes#6] [General]
Overall, I love the conciseness and the linear narrative. I know that there's many different ways to do the same thing with ggplot/ggproto and everyone in this space are very opinionated (me included - hence the "Opinionated" tags on some points below), so I appreciate that the reader is not bombarded with all the different choices in syntax, style, etc.. Also, the callouts are so good!

Here are some thoughts on my first read through. Please feel free to ignore any points that come across awkward (I didn't spend too long thinking about each):

- I think I would've liked to see a bit more of a "cake first" approach. So at the start where it currently says:

  > In this exercise, we’ll think about a way to add a point at the medians x and y, defining the new extension function geom_medians() and then you’ll be prompted to define geom_means() based on what you’ve learned.

  This could be accompanied with a visual example. For example, right below that you could also maybe show the desired syntax + the desired plot. This would also anchor the rest of the learning experience as "_the_ one thing" that the recipe is working up to. Then, the transition to Step 0 can be "... and here's how we _would have_ done that if we're restricted to using existing geoms"
  
- When you `penguins <- remove_missing(penguins)`, may be more palatable to assign it to something _new_ (`penguins_cleaned`) and printing something out for quick inspection (ex: via `glimpse()`)

- Maybe you could render the plot in Step 0 with the `geom_medians()` with `color = red`? Mainly for visually distinguishing the contribution of the geom but also to say "hey, it's just another geom that you can pass familiar params to!"

- (Opinionated) I know you've also discussed this with Teun elsewhere, but in Step 1 on the syntax of `function(data, scales)`, just to raise an alternative way of weaseling out of that potential confusion with introducing `scales` would be to define the function arguments as `(data, ...)` and inside the callout you can say "BTW other things get passed around the ggplot internals alongside the data, like `scales`, `params`, and so on. We don't do anything with them though here (you don't need to reach for those unless you're writing a more complicated extensions), so we simply let them through the dots `...` and we just focus on the `data`.)

- (Opinionated) This is beyond the scope of a recipe (which should just tell you what to do), but  in case you want to later expand on the discussion of naming conventions (which you currently have in Step 2 and elsewhere), I wrote a bit on that in a callout of my own! Under "On notations and conventions" in https://yjunechoe.github.io/posts/2022-07-06-ggplot2-delayed-aes-2/#ggproto-minus-the-gg-and-the-proto (feel free to take as you need)

- (Opinionated) On step 2, I think the `ggproto` with each argument on its own line is a perfect place to try out [Quarto code annotations](https://quarto.org/docs/authoring/code-annotation.html) (which may be worth applying elsewhere, if it turns out good)

- (Opinionated) Also on Step 2, one thing I like to do for demonstrating inheritance (at a broad, conceptual level) is to have people see how `Stat$compute_group` is defined (it's just a placeholder!) and use that to motivate the fact that our subclass extends the `$compute_group()` method specifically ("we take the template (`Stat`) and extend it where it tells us to (`$compute_group`)!"). Oversimplistic perhaps, but a good first mental model

- On Step 3, at the top, seeing the big fat constructor function first feels a bit intimidating. I agree with your advice to consule `stat_identity()` first, and maybe for this section you could gently introduce the body of how `stat_identity()` is written first (like just print `stat_identity`), then define `stat_medians()`, so readers can look back at `stat_identity()` and think (oh that wasn't so bad - we basically copy-pasted and swapped some default values)

- (Opinionated) On Step 3, when defining `geom_median()`, it may be kinda odd to let the layer's geom be swappable in a `geom_median(geom = ...)`. Not really a constructive feedback but something to keep in mind (I doubt anyone will actually try to pass a `geom` to a `geom_*()` function)

- On "Your Turn", I think this could use a whole lot more templating, especially given that the goal is to know your way around a recipe, not write/remember from scratch.

**Comment by EvaMaeRey** [comment]
Thank you @yjunechoe!!  responding point by point.  Hope to actually incorporate soon.

> Overall, I love the conciseness and the linear narrative. I know that there's many different ways to do the same thing with ggplot/ggproto and everyone in this space are very opinionated (me included - hence the "Opinionated" tags on some points below), so I appreciate that the reader is not bombarded with all the different choices in syntax, style, etc.. Also, the callouts are so good!

yay thanks.  Yeah, it seems like callouts are an opportunity to move things along without actually having to leave thing out.   

> Here are some thoughts on my first read through. Please feel free to ignore any points that come across awkward (I didn't spend too long thinking about each):

> * I think I would've liked to see a bit more of a "cake first" approach. So at the start where it currently says: "In this exercise, we’ll think about a way to add a point at the medians x and y, defining the new extension function geom_medians() and then you’ll be prompted to define geom_means() based on what you’ve learned."



>   This could be accompanied with a visual example. For example, right below that you could also maybe show the desired syntax + the desired plot. This would also anchor the rest of the learning experience as "_the_ one thing" that the recipe is working up to. Then, the transition to Step 0 can be "... and here's how we _would have_ done that if we're restricted to using existing geoms"


This is an important reminder.  For me, the syntax is cake!  But I think you have to be pretty far along the ggplot2 journey for that to be the case.  Reminds me of 'ggplot2: come for the beautiful plots, stay for the conceptual framework' 

> * When you `penguins <- remove_missing(penguins)`, may be more palatable to assign it to something _new_ (`penguins_cleaned`) and printing something out for quick inspection (ex: via `glimpse()`)

yup.

> * Maybe you could render the plot in Step 0 with the `geom_medians()` with `color = red`? Mainly for visually distinguishing the contribution of the geom but also to say "hey, it's just another geom that you can pass familiar params to!"

yup.

> * (Opinionated) I know you've also discussed this with Teun elsewhere, but in Step 1 on the syntax of `function(data, scales)`, just to raise an alternative way of weaseling out of that potential confusion with introducing `scales` would be to define the function arguments as `(data, ...)` and inside the callout you can say "BTW other things get passed around the ggplot internals alongside the data, like `scales`, `params`, and so on. We don't do anything with them though here (you don't need to reach for those unless you're writing a more complicated extensions), so we simply let them through the dots `...` and we just focus on the `data`.)

I like this idea.  I think it may take me a while to get used to.  But yeah, it would make compute_layer feel more similar to compute_group & compute_panel.   

> * (Opinionated) This is beyond the scope of a recipe (which should just tell you what to do), but  in case you want to later expand on the discussion of naming conventions (which you currently have in Step 2 and elsewhere), I wrote a bit on that in a callout of my own! Under "On notations and conventions" in https://yjunechoe.github.io/posts/2022-07-06-ggplot2-delayed-aes-2/#ggproto-minus-the-gg-and-the-proto (feel free to take as you need)

Thanks for the reference - I'm not sure that I dug into that callout.  Maybe I can start by linking there... 

> * (Opinionated) On step 2, I think the `ggproto` with each argument on its own line is a perfect place to try out [Quarto code annotations](https://quarto.org/docs/authoring/code-annotation.html) (which may be worth applying elsewhere, if it turns out good)

I like this idea, and hadn't noticed code annotation stuff.  Especially because this approach tries to keep the ggproto step really digestible w/ compute defined ahead of time, I think it is a good match - 'perfect place to try'!

> * (Opinionated) Also on Step 2, one thing I like to do for demonstrating inheritance (at a broad, conceptual level) is to have people see how `Stat$compute_group` is defined (it's just a placeholder!) and use that to motivate the fact that our subclass extends the `$compute_group()` method specifically ("we take the template (`Stat`) and extend it where it tells us to (`$compute_group`)!"). Oversimplistic perhaps, but a good first mental model

Okay, I think it's probably worth a call out, maybe I'll ask for help writing, especially if you have found this resonates and illuminates. 

> * On Step 3, at the top, seeing the big fat constructor function first feels a bit intimidating. I agree with your advice to consule `stat_identity()` first, and maybe for this section you could gently introduce the body of how `stat_identity()` is written first (like just print `stat_identity`), then define `stat_medians()`, so readers can look back at `stat_identity()` and think (oh that wasn't so bad - we basically copy-pasted and swapped some default values)

I'll think about a reversal of material.  looking at the stat_identity definition is totally new idea and it may well sit better before the 'doing'.  Or maybe at least that tip should be default collapse = F. My process for building the 'intimidating constructor function' has been mostly messy - actually go to a version of easy recipes and copy and paste, but really takes you out of flow.  In general I think this part of extension feels tricky, I think because it is mostly passive - there's a lot of syntax (things to get wrong?) but not necessarily a lot of decisions declared.


> * (Opinionated) On Step 3, when defining `geom_median()`, it may be kinda odd to let the layer's geom be swappable in a `geom_median(geom = ...)`. Not really a constructive feedback but something to keep in mind (I doubt anyone will actually try to pass a `geom` to a `geom_*()` function)

Alliasing like this just seems so convenient, and  not too likely to cause problems.  Else you are writing as much code as you did for stat_*?  i feel like the trade off is worth it.   Deliver on the promise of geom_* user facing function (cause people are comfortable with that) but also sneak in the stat_*() which is probably the more correct and truely useful layer to more fluent ggplot2 users.

> * On "Your Turn", I think this could use a whole lot more templating, especially given that the goal is to know your way around a recipe, not write/remember from scratch.

I felt that, and added the 'Done. Time to  review' code for proximate reference.  I guess I don't know if what templating would look like that would require engagement.  Did you have more specific thoughts?

**Reply by yjunechoe** [reply]
Thanks for engaging with me on all these points! I agree with your responses here - you've clearly already thought a lot about these things!

On the last point on the "Your Turn" section, I just want to explain what I meant more visually - I was specifically thinking of something more like a "fill in the blank" kind of a setup. I'm partly inspired by the new [Quarto exercises](https://r-wasm.github.io/quarto-live/exercises/exercises.html) feature they revealed recently at positconf. I don't think you _need_ a thorough hints and solutions setup here (though it'd be really cool to have!), but I vibe with the style of using `______` placeholders to focus on just the most important material.

So I'm imagining the "Your Turn" to have some parts pre-filled like this, just to remove some of the tediousness that's inherent to writing extensions:

![image](https://github.com/user-attachments/assets/ea015e6c-4513-4a09-8576-1de140aaadf2)

**Reply by EvaMaeRey** [reply]
I don't know.  I'm not sure the target audience, (at this point very proficient R and ggplot2 users), would benefit from this or find this an attractive interface.  I think I want to message that Stats (if not user-facing function) are pretty buildable from scratch, with a little practice.  Probably worth getting some data about - a poll or at least directing people to this specific question.  I do like your language: 'remove some of the tediousness that's inherent to writing extensions'!

**Reply by yjunechoe** [reply]
Ah yes - good point about the audience! I personally always appreciate hand-holding but I can imagine this being off-putting for users who are otherwise experts in R 😅. I think a poll / "how long did it take to complete this section" would be really helpful!

**Reply by EvaMaeRey** [reply]
I changed my mind!  I was probably just out of steam at the end of a longish list of items, motivating my arguments.  https://github.com/EvaMaeRey/easy-geom-recipes/blob/d77acd14a7648ad31f36c572be048099834af90f/recipe1means.qmd#L425  I think there are few folks need to be able to build from scratch. 

And in working it out, I think the templating emphasizes the cascade that you should come away with:  your base ggplot2 + data manipulation knowledge feeds to defining compute, compute defines Stat, Stat defines user-facer.

**Reply by yjunechoe** [reply]
Yay! Thanks for keeping the idea in mind and coming around to it :D

I think this will be so helpful - there are people who go through the _entire_ ggplot book and still skip the ggproto chapters because the book doesn't do the kind of handholding that I think are important!

**Reply by EvaMaeRey** [reply]
The one doubt I had was about running code that doesn't produce any output.  It feel like the webr chunks don't give you any kind of feedback, so then I put definition and tests together, so you get feedback when running the whole chunk.  But I think in separate chunks (definition separate from test) is nicer.  Do you have any insights as to a metaphorical 'ding' when a definition is executed.  Maybe I should assume something like progress bar will be added, and go with distinct define and test chunks.

**Reply by EvaMaeRey** [reply]
I'm thinking of rolling back on the `(data, ...)` to `(data, scales)` in the compute function definition.  I'm thinking about adding an argument to the compute definition, and I want it to be *one step away* from where we've been. So from knowing `(data, scales)` move to `(data, scales, myarg)`  I don't think you can go direct from (data, ...) to something that just works and that is 'one step away'.  Do you get what I mean?  Thoughts?

**Reply by EvaMaeRey** [reply]
I've just reread ggplot2 internals chapter - which has reference to Alice in Wonderland.  I think Hansel and Gretel is obvious choice for `...` discussion!  

> The next two sections follow the data down the rendering rabbit hole through the build step ([Section 19.2](https://ggplot2-book.org/internals#sec-ggplotbuild)) and the gtable step ([Section 19.3](https://ggplot2-book.org/internals#sec-ggplotgtable)) whereupon – rather like Alice in Lewis Carroll’s novel – it finally arrives in the grid wonderland as a collection of graphical primitives.

`grid wonderland!`

**Reply by yjunechoe** [reply]
Re: `(data, ...)` and `(data, scales)`, I get what you're saying there and I agree with you that having _just_ `data` would feel one-step gratuitously removed when it's both `data` _and_ `scales` that users would encounter when writing their own `$compute_group()` methods.

I was just thinking a little ahead to when they'd be writing other ggproto methods, after having learned that `data` and `scales` go together. Because in the grand scheme of things, most Stat methods that take a `data` argument don't take also `scale` argument, including `$compute_layer()`:

```r
Filter(\(x) is.function(x) && "data" %in% formalArgs(x), as.list(ggplot2::Stat)) |> 
  sapply(\(method) "scales" %in% formalArgs(method))
#> compute_layer    setup_data  finish_layer compute_panel compute_group  setup_params 
#>         FALSE         FALSE         FALSE          TRUE          TRUE         FALSE
```

In the case of Geom, _none_ of its methods that take `data` take a `scale` argument:

```r
Filter(\(x) is.function(x) && "data" %in% formalArgs(x), as.list(ggplot2::Geom)) |> 
  sapply(\(method) "scales" %in% formalArgs(method))
#> use_defaults   setup_data     draw_key   draw_group   draw_panel setup_params    handle_na   draw_layer 
#>        FALSE        FALSE        FALSE        FALSE        FALSE        FALSE        FALSE        FALSE
```

So I guess there's just a trade off of "how much do you want to reveal to users up front?" and "how much do you want users to unlearn later on?" - I don't have a good answer to this but just something to think about!

**Reply by yjunechoe** [reply]
Oh and to speak to the technical point about adding an argument, you can always do `(data, ..., myarg)`, and maintain the (intentional) "mysteriousness" of the dots as "internal details you don't have to worry about yet". I guess a bit more foreign and uglier, but again, a tradeoff 😓

**Reply by EvaMaeRey** [reply]
I was actually worried about some failure in passing the argument that I thought I was observing, but I don't see i now in this little test.  I should have done a reprex right then (I'm not even sure in what context - maybe lm w/ arbitrary aes), and at least before bringing up a role back.  I did think that going `scales` free was a good idea because the eventual interest in compute_layer, but your attention to the geoms has me team `...` again.  

``` r
library(tidyverse)
compute_group_summary_twoway <- function(data, ..., fun = mean){
  
  data %>% summarise(x = fun(x), y = fun(y))
  
}


StatSummaryTwoway <- ggproto("StatSummaryTwoway",
                             Stat,
                             compute_group = compute_group_summary_twoway
                             )

ggplot(cars) + 
  aes(speed, dist) + 
  geom_point() +
  geom_point(stat = StatSummaryTwoway, size = 8) + 
  aes(color = speed > 15)
```

![](https://i.imgur.com/Oh1UuNW.png)

``` r

ggplot(cars) + 
  aes(speed, dist) + 
  geom_point() + 
  geom_point(stat = StatSummaryTwoway, size = 8, fun = min) + 
  aes(color = speed > 15)
```

![](https://i.imgur.com/HC1IvlG.png)

``` r

compute_group_summary_twoway2 <- function(data, fun = mean, ...){
  
  data %>% summarise(x = fun(x), y = fun(y))
  
}


StatSummaryTwoway2 <- 
  ggproto("StatSummaryTwoway2",
          Stat,
          compute_group = compute_group_summary_twoway2)

ggplot(cars) + 
  aes(speed, dist) + 
  geom_point() + 
  geom_point(stat = StatSummaryTwoway2, size = 8) + 
  aes(color = speed > 15)
```

![](https://i.imgur.com/ZI3Ea5d.png)

``` r

ggplot(cars) + 
  aes(speed, dist) + 
  geom_point() + 
  geom_point(stat = StatSummaryTwoway2, size = 8, fun = min) + 
  aes(color = speed > 15)
```

![](https://i.imgur.com/CxqzTjj.png)

<sup>Created on 2024-09-10 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
I'm generally just really impressed all that time with your seeing, and even advocating the other side, @yjunechoe.  🏅 🎈

**Reply by EvaMaeRey** [reply]
> (Opinionated) On Step 3, when defining geom_median(), it may be kinda odd to let the layer's geom be swappable in a geom_median(geom = ...). Not really a constructive feedback but something to keep in mind (I doubt anyone will actually try to pass a geom to a geom_*() function)

Kind of piggybacking on some ggsprings stuff, I added the callout below to the `geom_medians <- stat_medians` verbatim alias move, which maybe just situates the move a bit more in a larger context of practice and gives people the model for what is more proper (but feels a bit forced, see the conclusion: 'However, because it is unexpected to use `geom_medians()` with a Stat other than StatMedians -- as doing so would remove the median-ness --- we think that the verbatim aliasing is a reasonable, time and code saving approach in cases like these.')  🤷‍♀️
. 


```
::: {.callout-warning collapse="true"}
# Be aware ...

... verbatim aliasing as shown above is a bit of a shortcut and assumes that users will use the 'geom_\*()' function with the stat-geom combination as-is. 

An approach that is more consistent with existing guidance would be to hardcode the Geom and allow the user to change the Stat as follows. (For a discussion, see [**Constructors**](https://ggplot2-book.org/ext-springs.html#constructors) in 'Extending ggplot2: A case Study' in *ggplot2: Elegant Graphics for Data Analysis*.  This section notes, 'Most ggplot2 users are accustomed to adding geoms, not stats, when building up a plot.') 


```{r}
# user-facing function
geom_medians <- function(mapping = NULL, data = NULL, 
                         stat = "medians", position = "identity", 
                         ..., show.legend = NA, inherit.aes = TRUE) 
{
    layer(data = data, mapping = mapping, stat = stat, 
        geom = GeomPoint, position = position, show.legend = show.legend, 
        inherit.aes = inherit.aes, params = rlang::list2(na.rm = FALSE, 
            ...))
}
```

However, because it is unexpected to use `geom_medians()` with a Stat other than StatMedians -- as doing so would remove the median-ness --- we think that the verbatim aliasing is a reasonable, time and code saving approach in cases like these.  
:::

```

**Comment by EvaMaeRey** [comment]
> (Opinionated) On Step 3, when defining geom_median(), it may be kinda odd to let the layer's geom be swappable in a geom_median(geom = ...). 

There concern here is with aliasing as follows

``` r 
geom_means <- stat_means

```


But maybe a move like the one below would make for more unsurprising interface (geom_*s Geoms are fixed, and Stats that are changable, and stat_*s where Stat is fixed and Geom is swappable) but allow for concisely-defined dedicated layers based on the same Stat...

``` r
layer_smooth_fit <- function (mapping = NULL, data = NULL, stat = StatSmoothFit, geom = GeomPointAccent, position = "identity", 
    ..., show.legend = NA, inherit.aes = TRUE) 
{
    layer(data = data, mapping = mapping, stat = stat, 
        geom = geom, position = position, show.legend = show.legend, 
        inherit.aes = inherit.aes, params = rlang::list2(na.rm = FALSE, 
            ...))
}

#' @export
stat_smooth_fit <- function(...){layer_smooth_fit(stat = StatSmoothFit, ...)}

#' @export
geom_smooth_fit <- function(...){layer_smooth_fit(geom = GeomPointAccent, ...)}

#' @export
geom_residuals <- function(...){layer_smooth_fit(geom = GeomSegmentAccent, ...)}

#' @export
geom_residual_springs <- function(...){layer_smooth_fit(geom = GeomSpringAccent, ...)}
```

Then you have the 'poems':

``` r
ggplot(cars) + 
   aes(dist, speed) + 
   geom_point() + 
   geom_smooth() + 
   geom_smooth_fit() + 
   geom_residuals()

ggplot(cars) + 
   aes(dist, speed) + 
   geom_point() + 
   geom_smooth() + 
   geom_smooth_fit() + 
   geom_residual_springs()
```

**Reply by EvaMaeRey** [reply]
make_constructor() should solve this problem (conciseness, conformity, and reproducing all arguments)!
